numpy/core/_internal.py

```diff
@@ -631,4 +631,17 @@ class TooHardError(RuntimeError):
     pass
 
 class AxisError(ValueError, IndexError):
-    pass
+    def __init__(self, axis, ndim=None, msg_prefix=None):
+        # single-argument form just delegates to base class
+        if ndim is None and msg_prefix is None:
+            msg = axis
+
+        # do the string formatting here, to save work in the C code
+        else:
+            msg = "axis {} is out of bounds for array of dimension {}".format(
```

Comment:
or `f"axis {axis} is out of bounds for array of dimension {ndim}"` and force everyone to use Python 3.6. 😈